tt_metal/impl/debug/dprint_server.cpp

```diff
@@ -411,6 +411,41 @@ void DebugPrintServerContext::DetachDevice(Device* device) {
         if (std::find(chip_ids.begin(), chip_ids.end(), device->id()) == chip_ids.end())
             return;
 
+    // When we detach a device, we should poll to make sure there's no outstanding prints.
+    chip_id_t chip_id = device->id();
+    uint32_t risc_mask = tt::llrt::OptionsG.get_dprint_riscv_mask();
+    bool outstanding_prints = true;
+    while (outstanding_prints && !server_killed_due_to_hang_) {
+        // Polling interval of 1ms
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
+
+        // Check all dprint-enabled cores on this device for outstanding prints.
+        outstanding_prints = false;
+        for (auto &core : device_to_core_range_.at(device)) {
+            for (int risc_id = 0; risc_id < GetNumRiscs(chip_id, core); risc_id++) {
+                if (risc_mask & (1<<risc_id)) {
+                    // No need to check if risc is not dprint-enabled.
+                    if (!CheckInitMagicCleared(chip_id, core, risc_id))
+                    continue;
+
+                    // Check if rpos < wpos, indicating unprocessed prints.
+                    constexpr int eightbytes = 8;
+                    uint32_t base_addr = GetBaseAddr(chip_id, core, risc_id);
+                    auto from_dev = tt::llrt::read_hex_vec_from_core(chip_id, core, base_addr, eightbytes);
```

Comment:
seems this races against the dprint_server?  if outstanding_prints is set, then we'll just wait.  if it is cleared...maybe the dprint_server read that there is a print but hasn't printed it yet?  or is there a lock so that if it gets that far then it's guaranteed to come out?

Reply:
I think we should be ok here - when the dprint server reads (rpos < wpos) and starts processing prints, it doesn't write rpos back until it after processes all prints for that risc. So I don't think we can have an issue with host thread reading (rpos == wpos) but the print server still hasn't processed the corresponding print data.

Let me know if I'm misunderstanding the question here though

Reply:
no, that covers it